scalanet/src/io/iohk/scalanet/peergroup/kademlia/KNetwork.scala

```diff
@@ -37,52 +54,82 @@ object KNetwork {
+      requestTemplate(to, request, { case p @ Pong(_, _) => p })
+    }
+
+    private def requestTemplate[Request <: KRequest[A], Response <: KResponse[A]](
+        to: NodeRecord[A],
+        message: Request,
+        pf: PartialFunction[KMessage[A], Response]
+    ): Task[Response] = {
       peerGroup
         .client(to.routingAddress)
         .bracket { clientChannel =>
-          makeFindNodesRequest(message, clientChannel)
+          sendRequest(message, clientChannel, pf)
         } { clientChannel =>
           clientChannel.close()
         }
-    }
 
-    override def findNodes: Observable[(FindNodes[A], Nodes[A] => Task[Unit])] = {
-      peerGroup.server().collectChannelCreated.mapTask { channel: Channel[A, KMessage[A]] =>
-        channel.in
-          .collect {
-            case f @ FindNodes(_, _, _) =>
-              (f, nodesTask(channel))
-          }
-          .headL
-          .timeout(requestTimeout)
-          .doOnFinish(closeIfAnError(channel))
-      }
     }
 
-    private def makeFindNodesRequest(message: FindNodes[A], clientChannel: Channel[A, KMessage[A]]): Task[Nodes[A]] = {
+    private def sendRequest[Request <: KRequest[A], Response <: KResponse[A]](
+        message: Request,
+        clientChannel: Channel[A, KMessage[A]],
+        pf: PartialFunction[KMessage[A], Response]
+    ): Task[Response] = {
       for {
         _ <- clientChannel.sendMessage(message).timeout(requestTimeout)
-        nodes <- clientChannel.in
-          .collect { case n @ Nodes(_, _, _) => n }
+        response <- clientChannel.in
+          .collect(pf)
           .headL
           .timeout(requestTimeout)
-      } yield nodes
+      } yield response
     }
 
-    private def closeIfAnError(
+    private def closeTheChannel[Request <: KRequest[A]](
         channel: Channel[A, KMessage[A]]
-    )(maybeError: Option[Throwable]): Task[Unit] = {
-      maybeError.fold(Task.unit)(_ => channel.close())
+    )(error: Throwable): (Request, KMessage[A] => Task[Unit]) = {
+      channel.close()
+      null
     }
 
-    private def nodesTask(
+    private def sendResponse(
         channel: Channel[A, KMessage[A]]
-    ): Nodes[A] => Task[Unit] = { nodes =>
+    ): KMessage[A] => Task[Unit] = { message =>
       channel
-        .sendMessage(nodes)
+        .sendMessage(message)
         .timeout(requestTimeout)
         .doOnFinish(_ => channel.close())
     }
+
+    private def serverTemplate[Request <: KRequest[A], Response <: KResponse[A]](
+        pf: PartialFunction[KMessage[A], Request]
+    ): Observable[(Request, KMessage[A] => Task[Unit])] = {
+      peerGroup
+        .server()
+        .collectChannelCreated
+        .mapEval { channel: Channel[A, KMessage[A]] =>
+          channel.in
+            .collect(pf)
+            .map(request => (request, sendResponse(channel)))
+            .headL
+            .timeout(requestTimeout)
+            .onErrorHandle(closeTheChannel(channel))
```

Comment:
Moreover, even if the consumer only wanted to handle one type of request (ping, say), it would still be necessary to close the channel for other request types (findNodes).

The only way I can see to do that would be for the calling code to pass some kind of message filter at construction time and the knetwork implementation to close the channel for everything not matching the filter.

That functionality is not required by KRouter (it wants to process everything), therefore I agree, a single message stream is a better approach.

Reply:
I've updated the code to provide a single request stream, as you suggest. This is still a little problematic because if the caller naively invokes `requests.filter(...)` or `requests.collect(...)` he introduces a resource leak. I've added a new class, `KNetworkExtension`, which (hopefully) provides methods that filter correctly.